src/core/blockchain/zilliqa/tokens/zrc2-client.ts

```diff
@@ -0,0 +1,65 @@
+import { Client } from '../client';
+import BigNumber from 'bignumber.js';
+import { fromBech32Address } from '@zilliqa-js/crypto';
+
+export class Zrc2Client {
+    constructor(private client: Client) {}
+
+    public async getBalance(contractAddress: string, accountAddress: string): Promise<BigNumber> {
+        const address = fromBech32Address(accountAddress).toLowerCase();
+
+        const smartContractSubState = await this.client.getSmartContractSubState(
+            contractAddress,
+            'balances',
+            [address]
+        );
+
+        const balance = (smartContractSubState?.balances || {})[address] || 0;
+        return new BigNumber(balance as string);
+    }
+
+    public async getSymbol(contractAddress) {
+        const smartContractInit = await this.getTokenDetails(contractAddress);
+        return this.findSmartContractSubField(smartContractInit, 'symbol');
+    }
+
+    public async getName(contractAddress) {
+        const smartContractInit = await this.getTokenDetails(contractAddress);
+        return this.findSmartContractSubField(smartContractInit, 'name');
+    }
+
+    public async getDecimals(contractAddress) {
+        const smartContractInit = await this.getTokenDetails(contractAddress);
+        return this.findSmartContractSubField(smartContractInit, 'decimals');
+    }
+
+    public async getTokenInfo(contractAddress) {
+        const info = await Promise.all([
```

Comment:
with this implementation you are doing 6 requests, but only 2 are needed. Merge this method with `getTokenDetails()` and in `getDecimals`, `getName `, `getSymbol` use that implementation. Example:
```ts
getTokenDetails(...) {
...
}

getDecimals(contractAddress) {
    let info = await this.getTokenDetails(contractAddress);
    return info.decimals;
}
```